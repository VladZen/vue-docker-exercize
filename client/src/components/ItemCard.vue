<template>
  <div
    class="inventory-item"
    :class="{
      shake: shake,
    }"
  >
    <div
      class="inventory-item__name"
      :class="{
        invalid: $errors.name,
      }"
    >
      <input
        ref="name-input"
        type="text"
        v-model="local.name"
        :disabled="mode === 'show'"
        :placeholder="mode === 'show' ? 'No Name' : 'Type name'"
      />

      <small class="error-msg">expects not to be empty!</small>
    </div>

    <div
      v-if="mode !== 'create'"
      class="inventory-item__quantity"
      :class="{
        invalid: $errors.quantity,
      }"
    >
      <input
        type="number"
        v-model.number="local.quantity"
        :disabled="mode === 'show'"
        placeholder="Type quantity"
      />

      <small class="error-msg">expects a positive integer!</small>
    </div>

    <div class="inventory-item__controls">
      <template v-if="mode === 'create'">
        <btn
          class="btn--save"
          title="Save"
          variant="success"
          :disabled="interactionInProgress"
          @click="save"
        >
          <spinner variant="success" :flag="!loaded.create" />
          <font-awesome-icon v-if="loaded.create" icon="check" />
        </btn>

        <btn
          class="btn--edit"
          title="Discard"
          variant="danger"
          :disabled="interactionInProgress"
          @click="reset"
        >
          <font-awesome-icon icon="undo" />
        </btn>
      </template>

      <template v-else-if="mode === 'show'">
        <btn
          class="btn--edit"
          title="Edit"
          :disabled="interactionInProgress"
          @click="toggleMode('edit')"
        >
          <font-awesome-icon icon="edit" />
        </btn>

        <btn
          class="btn--remove"
          title="Remove"
          variant="danger"
          :disabled="interactionInProgress"
          @click="remove"
        >
          <spinner variant="danger" :flag="!loaded.remove" />
          <font-awesome-icon v-if="loaded.remove" icon="times" />
        </btn>
      </template>

      <template v-else-if="mode === 'edit'">
        <btn
          v-if="mode === 'edit'"
          class="btn--apply"
          title="Apply"
          variant="success"
          :disabled="interactionInProgress"
          @click="update"
        >
          <spinner variant="success" :flag="!loaded.update" />
          <font-awesome-icon v-if="loaded.update" icon="check" />
        </btn>

        <btn
          key="discard"
          class="btn--edit"
          title="Discard"
          variant="danger"
          :disabled="interactionInProgress"
          @click="
            toggleMode('show');
            reset(item);
          "
        >
          <font-awesome-icon icon="undo" />
        </btn>
      </template>
    </div>
  </div>
</template>

<script>
import _isEqual from "lodash.isequal";
import defaultItem from "@/config/default-item.js";

import { reactive, ref, computed, watch } from "@vue/composition-api";
import useFlag from "@/use/requestLoadingFlag";
import useStore from "@/use/store";

import Spinner from "@/components/Spinner";
import Btn from "@/components/Button";

export default {
  components: {
    Spinner,
    Btn,
  },
  setup(props, ctx) {
    const store = useStore(ctx);
    const mode = ref(props.state);
    const local = ref(defaultItem(props.item));
    const $errors = reactive({ name: false, quantity: false });
    const shake = ref(false);

    const { loaded: create, promise: dispatchCreate } = useFlag(
      store.dispatch.bind(null, "ADD")
    );
    const { loaded: update, promise: dispatchUpdate } = useFlag(
      store.dispatch.bind(null, "UPDATE")
    );
    const { loaded: remove, promise: dispatchRemove } = useFlag(
      store.dispatch.bind(null, "REMOVE")
    );
    const loaded = reactive({
      create,
      update,
      remove,
    });
    const interactionInProgress = computed(() =>
      Object.keys(loaded).some((action) => !loaded[action])
    );

    watch(
      () => props.item,
      (value) => (local.value = { ...value })
    );

    return {
      loaded,
      dispatchCreate,
      dispatchUpdate,
      dispatchRemove,
      interactionInProgress,
      defaultItem: () => reactive(defaultItem(props.item)),
      local,
      mode,
      $errors,
      shake,
    };
  },
  props: {
    item: {
      type: Object,
      default: () => {},
    },
    state: {
      type: String,
      validator: (value) => ["show", "create", "edit"].includes(value),
      default: "show",
    },
  },
  methods: {
    async save() {
      if (this.isInvalid()) return;
      const id = await this.dispatchCreate(this.local.name);
      this.$emit("save", id);
      this.reset();
    },
    async remove() {
      await this.dispatchRemove(this.local);
      this.$emit("remove", this.local.id);
    },
    async update() {
      if (this.isInvalid()) return;

      if (!_isEqual(this.item, this.local)) {
        this.local = await this.dispatchUpdate(this.local);
        this.$emit("update", this.local.id);
      }
      this.toggleMode("show", this.local.id);
    },
    toggleMode(mode, id = this.local.id) {
      this.mode = mode;
      this.$emit("update:mode", { mode, id });
    },
    reset(to = null) {
      this.local = this.defaultItem(to);
      this.$errors = {
        name: false,
        quantity: false,
      };
      this.$emit("reset");
    },
    shakeItem() {
      this.shake = true;
      setTimeout(() => (this.shake = false), 1000);
    },
    isInvalid() {
      this.$errors.name = !this.local.name;
      if (this.mode === "edit")
        this.$errors.quantity =
          ["", null, undefined].includes(this.local.quantity) ||
          this.local.quantity < 0;

      const error = this.$errors.name || this.$errors.quantity;
      if (error) this.shakeItem();
      return error;
    },
  },
};
</script>

<style lang="scss">
.inventory-item {
  display: flex;
  align-items: flex-end;
  justify-content: space-between;
  background-color: #fff;
  border-radius: $base-radius;
  padding: 5px $base-offset * 2;
  min-height: 83px;

  &__quantity {
    margin-left: 10px;
    display: inline-flex;
    align-items: baseline;
    margin-right: $base-offset;

    input {
      text-align: right;
    }
  }

  &__quantity,
  &__name {
    display: flex;
    flex-direction: column;
  }

  &__name {
    flex: 2;
  }

  &__quantity {
    flex: 1;
  }

  &__controls {
    align-self: center;
  }

  &.shake {
    animation: shake 1s cubic-bezier(0.36, 0.07, 0.19, 0.97) both;
    transform: translate3d(0, 0, 0);
    backface-visibility: hidden;
    perspective: 1000px;
  }

  .error-msg {
    opacity: 0;
    color: $danger-color;
    transition: opacity $transition;
    pointer-events: none;
  }

  .invalid {
    input {
      border-color: $danger-color;
    }

    .error-msg {
      opacity: 1;
      pointer-events: all;
    }
  }

  & + & {
    margin-top: $base-offset;
  }
}

@keyframes shake {
  10%,
  90% {
    transform: translate3d(-1px, 0, 0);
  }

  20%,
  80% {
    transform: translate3d(2px, 0, 0);
  }

  30%,
  50%,
  70% {
    transform: translate3d(-4px, 0, 0);
  }

  40%,
  60% {
    transform: translate3d(4px, 0, 0);
  }
}
</style>
