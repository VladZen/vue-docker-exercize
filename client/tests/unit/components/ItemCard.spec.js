import { mount, createLocalVue } from "@vue/test-utils";
import ItemCard from "@/components/ItemCard";
import Vuex from "vuex";
import defaultItem from "@/config/default-item";

const localVue = createLocalVue();
localVue.use(Vuex);

const SAVE_ID = "1";
const SAVE_NAME = "New Item";
const ABSTRACT_QUANTITY = 30;
const EXISTING_ITEM = defaultItem({
  id: SAVE_ID,
  name: SAVE_NAME,
  quantity: ABSTRACT_QUANTITY,
});

const store = new Vuex.Store({
  actions: {
    ADD: () => SAVE_ID,
    UPDATE: (store, payload) => payload,
    REMOVE: (store, { id }) => id,
  },
});

const mountItemCard = (propsData = {}) => {
  const div = document.createElement("div");
  document.body.appendChild(div);

  return mount(ItemCard, {
    store,
    localVue,
    propsData,
    attachTo: div,
    stubs: {
      "font-awesome-icon": true,
    },
  });
};

describe("ItemCard", () => {
  let wrapper;
  const getButtons = () => wrapper.findAll(".inventory-item__controls .btn");

  it.each(["show", "create", "edit"])(
    'checks "%s" basic rendering',
    (state) => {
      if (["show", "edit"].includes(state)) {
        expect(mountItemCard({ state, item: EXISTING_ITEM })).toMatchSnapshot();
      } else {
        expect(mountItemCard({ state: "create" })).toMatchSnapshot();
      }
    }
  );

  describe("create", () => {
    beforeEach(() => (wrapper = mountItemCard({ state: "create" })));
    afterEach(() => wrapper.destroy());

    describe("should properly handle save", () => {
      it("when input data is invalid", async () => {
        const btn = getButtons().at(0);
        await wrapper.setData({
          local: {
            name: "",
          },
        });

        await btn.trigger("click");
        expect(wrapper.emitted().save).toBeFalsy();
        expect(wrapper).toMatchSnapshot();
      });

      it("when input data is valid", async () => {
        const btn = getButtons().at(0);
        await wrapper.setData({
          local: {
            name: SAVE_NAME,
          },
        });

        await btn.trigger("click");
        await global.timeout(1);
        expect(wrapper.emitted().save[0][0]).toBe(SAVE_ID);
      });
    });

    it("should properly handle discard", async () => {
      const btn = getButtons().at(1);
      await btn.trigger("click");
      expect(wrapper.emitted().reset).toBeTruthy();
    });
  });

  describe("edit", () => {
    beforeEach(
      () => (wrapper = mountItemCard({ state: "edit", item: EXISTING_ITEM }))
    );
    afterEach(() => wrapper.destroy());

    describe("should properly handle update", () => {
      const btn = () => getButtons().at(0);

      it("when input data is invalid", async () => {
        await wrapper.setData({
          local: {
            name: "",
            quantity: "",
          },
        });

        await btn().trigger("click");
        expect(wrapper.emitted().update).toBeFalsy();
        expect(wrapper.emitted()["update:mode"]).toBeFalsy();
        expect(wrapper).toMatchSnapshot();
      });

      it("when data has not been changed", async () => {
        await btn().trigger("click");
        expect(wrapper.emitted().update).toBeFalsy();
        expect(wrapper.emitted()["update:mode"][0][0]).toEqual({
          mode: "show",
          id: SAVE_ID,
        });
        expect(wrapper).toMatchSnapshot();
      });

      it("when data has been changed and valid", async () => {
        const newData = {
          id: SAVE_ID,
          name: "Item Changed",
          quantity: 50,
        };

        await wrapper.setData({ local: newData });

        await btn().trigger("click");
        await global.timeout(1);
        expect(wrapper.emitted().update[0][0]).toBe(SAVE_ID);
        expect(wrapper.emitted()["update:mode"][0][0]).toEqual({
          mode: "show",
          id: SAVE_ID,
        });
        expect(wrapper).toMatchSnapshot();
      });
    });

    it("should properly handle discard", async () => {
      const btn = getButtons().at(1);
      const newData = {
        id: SAVE_ID,
        name: "Item Changed",
        quantity: 50,
      };
      await wrapper.setData({ local: newData });
      await btn.trigger("click");
      expect(wrapper.emitted()["update:mode"][0][0]).toEqual({
        mode: "show",
        id: SAVE_ID,
      });
      expect(wrapper.emitted().reset).toBeTruthy();
      Object.keys(wrapper.vm.$errors).forEach((key) => {
        expect(wrapper.vm.$errors[key]).toBeFalsy();
      });
      expect(wrapper.vm.local).toEqual(EXISTING_ITEM);
    });
  });

  describe("show", () => {
    beforeEach(
      () => (wrapper = mountItemCard({ state: "show", item: EXISTING_ITEM }))
    );
    afterEach(() => wrapper.destroy());

    it("should properly handle changing mode to edit", async () => {
      const btn = getButtons().at(0);
      await btn.trigger("click");
      expect(wrapper.emitted()["update:mode"][0][0]).toEqual({
        mode: "edit",
        id: SAVE_ID,
      });
    });

    it("should properly handle remove", async () => {
      const btn = getButtons().at(1);
      await btn.trigger("click");
      await global.timeout(1);
      expect(wrapper.emitted().remove[0][0]).toBe(SAVE_ID);
    });
  });
});
