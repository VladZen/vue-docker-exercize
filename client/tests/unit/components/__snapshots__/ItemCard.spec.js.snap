// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemCard checks "create" basic rendering 1`] = `
<form class="inventory-item">
  <div class="inventory-item__name"><input type="text" placeholder="Type name"> <small class="error-msg">expects not to be empty!</small></div>
  <!---->
  <div class="inventory-item__controls"><button type="submit" class="btn btn--save btn--success" title="Save">
      <!---->
      <font-awesome-icon-stub icon="check"></font-awesome-icon-stub>
    </button> <button type="reset" class="btn btn--edit btn--danger" title="Discard">
      <font-awesome-icon-stub icon="undo"></font-awesome-icon-stub>
    </button></div>
</form>
`;

exports[`ItemCard checks "edit" basic rendering 1`] = `
<form class="inventory-item">
  <div class="inventory-item__name"><input type="text" placeholder="Type name"> <small class="error-msg">expects not to be empty!</small></div>
  <div class="inventory-item__quantity"><input type="number" placeholder="Type quantity"> <small class="error-msg">expects a positive integer!</small></div>
  <div class="inventory-item__controls"><button type="submit" class="btn btn--apply btn--success" title="Apply">
      <!---->
      <font-awesome-icon-stub icon="check"></font-awesome-icon-stub>
    </button> <button type="reset" class="btn btn--edit btn--danger" title="Discard">
      <font-awesome-icon-stub icon="undo"></font-awesome-icon-stub>
    </button></div>
</form>
`;

exports[`ItemCard checks "show" basic rendering 1`] = `
<form class="inventory-item">
  <div class="inventory-item__name"><input type="text" disabled="disabled" placeholder="No Name"> <small class="error-msg">expects not to be empty!</small></div>
  <div class="inventory-item__quantity"><input type="number" disabled="disabled" placeholder="Type quantity"> <small class="error-msg">expects a positive integer!</small></div>
  <div class="inventory-item__controls"><button type="button" class="btn btn--edit" title="Edit">
      <font-awesome-icon-stub icon="edit"></font-awesome-icon-stub>
    </button> <button type="button" class="btn btn--remove btn--danger" title="Remove">
      <!---->
      <font-awesome-icon-stub icon="times"></font-awesome-icon-stub>
    </button></div>
</form>
`;

exports[`ItemCard create should properly handle save when input data is invalid 1`] = `
<form class="inventory-item shake">
  <div class="inventory-item__name invalid"><input type="text" placeholder="Type name"> <small class="error-msg">expects not to be empty!</small></div>
  <!---->
  <div class="inventory-item__controls"><button type="submit" class="btn btn--save btn--success" title="Save">
      <!---->
      <font-awesome-icon-stub icon="check"></font-awesome-icon-stub>
    </button> <button type="reset" class="btn btn--edit btn--danger" title="Discard">
      <font-awesome-icon-stub icon="undo"></font-awesome-icon-stub>
    </button></div>
</form>
`;

exports[`ItemCard edit should properly handle update when data has been changed and valid 1`] = `
<form class="inventory-item">
  <div class="inventory-item__name"><input type="text" placeholder="No Name" disabled="disabled"> <small class="error-msg">expects not to be empty!</small></div>
  <div class="inventory-item__quantity"><input type="number" placeholder="Type quantity" disabled="disabled"> <small class="error-msg">expects a positive integer!</small></div>
  <div class="inventory-item__controls"><button type="button" class="btn btn--edit" title="Edit">
      <font-awesome-icon-stub icon="edit"></font-awesome-icon-stub>
    </button> <button type="button" class="btn btn--remove btn--danger" title="Remove">
      <!---->
      <font-awesome-icon-stub icon="times"></font-awesome-icon-stub>
    </button></div>
</form>
`;

exports[`ItemCard edit should properly handle update when data has not been changed 1`] = `
<form class="inventory-item">
  <div class="inventory-item__name"><input type="text" placeholder="No Name" disabled="disabled"> <small class="error-msg">expects not to be empty!</small></div>
  <div class="inventory-item__quantity"><input type="number" placeholder="Type quantity" disabled="disabled"> <small class="error-msg">expects a positive integer!</small></div>
  <div class="inventory-item__controls"><button type="button" class="btn btn--edit" title="Edit">
      <font-awesome-icon-stub icon="edit"></font-awesome-icon-stub>
    </button> <button type="button" class="btn btn--remove btn--danger" title="Remove">
      <!---->
      <font-awesome-icon-stub icon="times"></font-awesome-icon-stub>
    </button></div>
</form>
`;

exports[`ItemCard edit should properly handle update when input data is invalid 1`] = `
<form class="inventory-item shake">
  <div class="inventory-item__name invalid"><input type="text" placeholder="Type name"> <small class="error-msg">expects not to be empty!</small></div>
  <div class="inventory-item__quantity invalid"><input type="number" placeholder="Type quantity"> <small class="error-msg">expects a positive integer!</small></div>
  <div class="inventory-item__controls"><button type="submit" class="btn btn--apply btn--success" title="Apply">
      <!---->
      <font-awesome-icon-stub icon="check"></font-awesome-icon-stub>
    </button> <button type="reset" class="btn btn--edit btn--danger" title="Discard">
      <font-awesome-icon-stub icon="undo"></font-awesome-icon-stub>
    </button></div>
</form>
`;
